game-app/game-core/src/main/java/games/strategy/triplea/ui/panels/map/MapPanel.java

```diff
@@ -584,14 +583,36 @@ private boolean unitsChanged(final Tuple<Territory, List<Unit>> newUnits) {
                 newUnits.getSecond(), currentUnits.getSecond()));
   }
 
-  public void updateCountries(final Collection<Territory> countries) {
-    tileManager.updateTerritories(countries, gameData, uiContext.getMapData());
-    smallMapImageManager.update(uiContext.getMapData());
-    SwingUtilities.invokeLater(
-        () -> {
-          smallView.repaint();
-          repaint();
-        });
+  public void updateCountries(Collection<Territory> countries) {
+    // When there are multiple updateCountries() notifications in a row, for example from going far
+    // back in history, no need to do this repeatedly. Instead, a single update is possible if the
+    // async code runs after all the notifications have been received.
+    final boolean scheduleUpdate;
+    synchronized (countriesToUpdate) {
+      scheduleUpdate = countriesToUpdate.isEmpty();
+      countriesToUpdate.addAll(countries);
+    }
+    if (!scheduleUpdate) {
+      return; // An update is already scheduled.
+    }
+    AsyncRunner.runAsync(
+            () -> {
+              Collection<Territory> toUpdate;
+              synchronized (countriesToUpdate) {
+                // Note: Don't run updateTerritories() inside countriesToUpdate lock, as this causes
+                // a deadlock due to locking game data inside updateTerritories().
+                toUpdate = List.copyOf(countriesToUpdate);
+                countriesToUpdate.clear();
```

Comment:
If you lock on some other object, you could just assign countriesToUpdate to the `toUpdate` variable and assign it a new instance of a fresh set implementation. This would avoid a copy operation.

Reply:
Done.